### PR TITLE
amazon: fix s3 client caching

### PR DESCRIPTION
**Pull request by jeffswenson**
This fixes an issue introduced by #132953. The issue only impacts the 23.2 back port. If the S3 client config specifies an explicit endpoint and it does not specify an explicit region, the client cache is thrashed. This cache thrashing was detected by the cloud unit tests when a worker pool change moved the tests farther away from the S3 bucket.

Fixes: #134021
Fixes: #133901

Release Note: None. Issue caught before release cut.

### CODE REVIEW

**Comment by jeffswenson** [comment]
https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/17638517

https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/17638515

**Review by msbutler — APPROVED** [review]
Just to double check, have you run the cloud nightly suite a few times on this patch?

**Comment by jeffswenson** [comment]
The cloud unit test suite is still failing even with this patch. Here's a run with the new team city config and an old build. It's possible this is a red-herring and isn't related to the test failures.

https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/17638936

**Comment by jeffswenson** [comment]
Okay, so it looks like the test failures are not actually caused by https://github.com/cockroachdb/cockroach/pull/132953. https://github.com/cockroachdb/cockroach/pull/134512#issuecomment-2462476559 fails with the same error and it was run against a commit that does not include the back port.

**Comment by jeffswenson** [comment]
TLDR: I think this is fixing a real bug that breaks the caching logic in niche use cases. But the bug is not actually the reason the cloud nightly unit tests are failing. The tests are just slow because the worker pool was moved far away from the destination. It kind of makes sense that this doesn't have a big impact on the test perforamnce. The cache is a singleton cache and is already easy to trash, so the tests likely thrash it regardless of the bug.
